graylog2-server/src/main/java/org/graylog2/indexer/messages/Messages.java

```diff
@@ -218,21 +228,85 @@ public ResultMessage get(String messageId, String index) throws DocumentNotFound
             if (LOG.isDebugEnabled()) {
                 String chunkInfo = "";
                 if (chunkSize != messageList.size()) {
-                    chunkInfo = String.format(Locale.ROOT, " (chunk %d/%d offset %d)", partitionCount,
-                            (int) Math.ceil((double)messageList.size() / chunkSize), offset);
+                    chunkInfo = String.format(Locale.ROOT, " (chunk %d/%d offset %d)", chunkCount,
+                            (int) Math.ceil((double) messageList.size() / chunkSize), offset);
                 }
                 LOG.debug("Index: Bulk indexed {} messages{}, failures: {}",
                         result.getItems().size(), chunkInfo, failedItems.size());
             }
-            if (!result.getFailedItems().isEmpty()) {
+            if (!remainingFailures.isEmpty()) {
                 LOG.error("Failed to index [{}] messages. Please check the index error log in your web interface for the reason. Error: {}",
-                        result.getFailedItems().size(), result.getErrorMessage());
+                        remainingFailures.size(), result.getErrorMessage());
             }
-            partitionCount++;
+            chunkCount++;
         }
         return failedItems;
     }
 
+    private BulkResult bulkIndexChunk(List<Map.Entry<IndexSet, Message>> chunk) {
+        Bulk.Builder bulk = new Bulk.Builder();
+
+        for (Map.Entry<IndexSet, Message> entry : chunk) {
+            final Message message = entry.getValue();
+
+            bulk.addAction(new Index.Builder(message.toElasticSearchObject(invalidTimestampMeter))
+                    .index(entry.getKey().getWriteIndexAlias())
+                    .type(IndexMapping.TYPE_MESSAGE)
+                    .id(message.getId())
+                    .build());
+        }
+
+        return runBulkRequest(bulk.build(), chunk.size());
+    }
+
+    private Set<BulkResult.BulkResultItem> retryOnlyIndexBlockItemsForever(List<Map.Entry<IndexSet, Message>> chunk, List<BulkResult.BulkResultItem> allFailedItems) {
+        Set<BulkResult.BulkResultItem> indexBlocks = indexBlocksFrom(allFailedItems);
+        Set<BulkResult.BulkResultItem> otherFailures = new HashSet<>(Sets.difference(new HashSet<>(allFailedItems), indexBlocks));
+        List<Map.Entry<IndexSet, Message>> blockedMessages = messagesForResultItems(chunk, indexBlocks);
+
+        long attempt = 1;
+
+        while (!indexBlocks.isEmpty()) {
+            blockExecutionForAttempt(attempt++);
+
+            BulkResult bulkResult = bulkIndexChunk(blockedMessages);
+
+            List<BulkResult.BulkResultItem> failedItems = bulkResult.getFailedItems();
+
+            indexBlocks = indexBlocksFrom(failedItems);
+            blockedMessages = messagesForResultItems(blockedMessages, indexBlocks);
+
+            Set<BulkResult.BulkResultItem> newOtherFailures = Sets.difference(new HashSet<>(failedItems), indexBlocks);
+            otherFailures.addAll(newOtherFailures);
+        }
+
```

Comment:
It would be good to have some logging here indicating that the index blocks have been resolved.

Reply:
:+1: Agreed. I added some.